### ISSUE TITLE
Indexes and constraints not getting restored correctly

### ISSUE THREAD

**Issue by AndreaCrotti** [dimitri/pgloader#400]
So I think today I finally understood how pgloader works, correct me if I'm wrong:

- it drops all the indexes/constraints
- it copies stuff in parallel
- it restores all checks again in parallel

Now I noticed an issue, which is that the indexes/constraints that get recreated are not defined exactly in the same way.

    Foreign-key constraints:
        "access_log__session_id_3dc2579f67feddc_fk_access_log_session_id" FOREIGN KEY (session_id) REFERENCES access_log_session(id) DEFERRABLE INITIALLY DEFERRED
        "access_log_request_site_id_680ca1605ee80e0a_fk_django_site_id" FOREIGN KEY (site_id) REFERENCES django_site(id) DEFERRABLE INITIALLY DEFERRED
    Referenced by:
        TABLE "access_log_activitywindow" CONSTRAINT "ac_closing_request_id_5c24c0371195dccc_fk_access_log_request_id" FOREIGN KEY (closing_request_id) REFERENCES access_log_request(id) DEFERRABLE INITIALLY DEFERRED
        TABLE "access_log_activitywindow" CONSTRAINT "ac_opening_request_id_2181f178d1c767d1_fk_access_log_request_id" FOREIGN KEY (opening_request_id) REFERENCES access_log_request(id) DEFERRABLE INITIALLY DEFERRED
        TABLE "access_log_impersonateinfo" CONSTRAINT "access_log__request_id_94714564cadfecd_fk_access_log_request_id" FOREIGN KEY (request_id) REFERENCES access_log_request(id) DEFERRABLE INITIALLY DEFERRED
        TABLE "access_log_utm" CONSTRAINT "access_log_request_id_5c577bd5f8aaf3c8_fk_access_log_request_id" FOREIGN KEY (request_id) REFERENCES access_log_request(id) DEFERRABLE INITIALLY DEFERRED
        TABLE "redirects_redirectinstance" CONSTRAINT "redirects__request_id_4ff7f017209814f1_fk_access_log_request_id" FOREIGN KEY (request_id) REFERENCES access_log_request(id) DEFERRABLE INITIALLY DEFERRED

And this is how they are defined when using pgloader:

    Foreign-key constraints:
        "session_id_refs_id_45743e5c" FOREIGN KEY (session_id) REFERENCES access_log_session(id) ON UPDATE RESTRICT ON DELETE RESTRICT
        "site_id_refs_id_1733372c" FOREIGN KEY (site_id) REFERENCES django_site(id) ON UPDATE RESTRICT ON DELETE RESTRICT
    Referenced by:
        TABLE "access_log_activitywindow" CONSTRAINT "closing_request_id_refs_id_b56bfd3a" FOREIGN KEY (closing_request_id) REFERENCES access_log_request(id) ON UPDATE RESTRICT ON DELETE RESTRICT
        TABLE "access_log_activitywindow" CONSTRAINT "opening_request_id_refs_id_b56bfd3a" FOREIGN KEY (opening_request_id) REFERENCES access_log_request(id) ON UPDATE RESTRICT ON DELETE RESTRICT
        TABLE "redirects_redirectinstance" CONSTRAINT "request_id_refs_id_1bc70ef5" FOREIGN KEY (request_id) REFERENCES access_log_request(id) ON UPDATE RESTRICT ON DELETE RESTRICT
        TABLE "access_log_impersonateinfo" CONSTRAINT "request_id_refs_id_3989d9bf" FOREIGN KEY (request_id) REFERENCES access_log_request(id) ON UPDATE RESTRICT ON DELETE RESTRICT
        TABLE "access_log_utm" CONSTRAINT "request_id_refs_id_aa6e5cea" FOREIGN KEY (request_id) REFERENCES access_log_request(id) ON UPDATE RESTRICT ON DELETE RESTRICT

Now I'm not sure if this is a real issue in terms of behaviour, but it would mean that all the local dev databases are not defined in the same way, and would really like to avoid that.

Is this known behaviour?
Is there a way to avoid this?

Thanks a lot

**Comment by dimitri** [comment]
What is your source database ? MySQL I believe ? pgloader doesn't know anything about `DEFERRABLE INITIALLY DEFERRED` yet, that's why it's not migrated over. We need to find out where to find the information in the catalogs and then add that knowledge to pgloader... See https://github.com/dimitri/pgloader/blob/master/src/sources/mysql/mysql-schema.lisp#L243 and following for some level of details.

**Comment by AndreaCrotti** [comment]
Yes it's MySQL, and we are migrating from MySQL in aws RDS to Postgres in aws RDS.

So and what about the "ON UPDATE RESTRICT" and "ON DELETE RESTRICT" that were added?

They don't seem to be the default option so I guess that would behave differently as well..
Anyway even assuming I fix these two, is there a way later on to compare the schema before and after and check that they are exactly the same with pgloader?

Or should I use an external tool for that?
thanks!

**Comment by AndreaCrotti** [comment]
Actually just to clarify in case it's not clear, we are migrating from MySQL but these indexes checks are defined in the destination database *before* I start the migration, since I used the schema generated by Django migrations.

So in theory (probably making it sound much easier than it actually is) pgloader should just re-add exactly the same indexes/constraints that it deleted..

**Comment by dimitri** [comment]
Oh, I see. Then play with the `WITH include no drop, create no indexes, no foreign keys` options, or even `WITH data only` as django should have created the whole schema already.

Performance wise, you should actually drop all the indexes and constraints before the load and recreate them once it's done. There's currently no option to tell pgloader to do a *data only* migration in which it would do that for you automatically, taking the catalog details from PostgreSQL rather than from your source database. That would be an interesting option, though.

**Comment by AndreaCrotti** [comment]
Ah ok thanks I'll try that.
The problem with dropping and readding indexes is that I don't really have any other tool to do that, and there are hundreds of tables to migrate..

There are only a handful of big tables though so maybe if I just take care of them I'm fine, but in any case since pgloader already does this but just taking the catalog from another db it might not be so hard, could you give me a few pointers about what I would need to modify?

Is pgloader currently able at least to read the catalog from Postgres instead of just writing to it or that is also missing?

Thanks

**Comment by dimitri** [comment]
We currently have parts of the logic to read from PostgreSQL catalogs (indexes), you can read parts of it in https://github.com/dimitri/pgloader/blob/master/src/pgsql/queries.lisp#L252 and following. I think it's time to consider PostgreSQL as a source of data though and implement a src/sources/pgsql directory hosting a pgsql-schema.lisp file following the example of https://github.com/dimitri/pgloader/blob/master/src/sources/mysql/mysql-schema.lisp. It should then allow to specify reading the catalogs from an existing PostgreSQL target database, and also to support PostgreSQL to PostgreSQL operations.

**Comment by AndreaCrotti** [comment]
Ah yes I think it would be good, however I think to actually fit in this use case I'm talking about (which I guess is not that uncommon), we would need to be able to get the data from one source, and get the catalog from another source (in this case the same Postgres database).
thanks

**Comment by dimitri** [comment]
Exactly, your case needs data from the source and catalogs from the target, so we need an option to do just that. What I'm referring to is code layout where I think the catalog reflexion queries should be developed as the `sources/pgsql/pgsql-schema.lisp` facility for later re-use in the support of PostgreSQL as a target.

**Comment by AndreaCrotti** [comment]
Hi @dimitri 

I'm still working on this issue and still haven't figured out a final solution.

Have to decide if trying to add this feature here or just do some magic with the schema generated by Django.

Another thing though, does pgloader supports some kind of streaming?
I can see in the docs that it's using the COPY, but I'm not sure it's the same thing.
  
I mean I could not just leave it running forever keeping two databases in sync, right?

Because sadly some tables are going to take ages to copy in any case, which means quite a lot of downtime unless I find a smarter way to do it.

Any idea how do people normally do big migrations?
Thanks

**Comment by dimitri** [comment]
I might get around to just implement the needed option here (catalogs from the target database and data from the source database).

In the context of pgloader `COPY` is about streaming a particular data set into PostgreSQL. More generally it can also be used the other way around, to extract a snapshot of the data as a stream.

For big migrations either you're in a position where you can accept the downtime required to copy the data from the source to the target instances, or you need some kind of replication. See https://github.com/dimitri/pgloader/issues/166 for details on how to implement that: I'm interested into doing it, would appreciate some sponsoring to put a good enough chunk of personal time on the topic!
